{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BF527: Lab 10\n",
    "\n",
    "## Rosalind Bioinformatics Stronghold\n",
    "\n",
    "The website http://rosalind.info has a number of valuable programming exercises to help teach core bioinformatics concepts. To give you practice with writing and running python scripts, a command line version of the Rosalind Stronghold problems has been created for you. Some labs will have a Rosalind problem for you to solve as part of the lab, and this is the first such lab.\n",
    "\n",
    "## Installing rosalind\n",
    "First let's install rosalind in an anaconda environment on SCC from <a hre=\"https://anaconda.org/bubhub/rosalind-stronghold\">BUBHUB</a>. \n",
    "\n",
    "`module load anaconda3`\n",
    "\n",
    "`conda create -n rosalind python=3`\n",
    "\n",
    "`source activate rosalind`\n",
    "\n",
    "`conda install -c bubhub rosalind-stronghold`\n",
    "\n",
    "\n",
    "## Running rosalind\n",
    "\n",
    "To get started, run the following on the command line:\n",
    "\n",
    "`rosalind help`\n",
    "\n",
    "The information printed to the screen will give you instructions on how to use the `rosalind` program.\n",
    "\n",
    "The rosalind problem ID you will solve today is **DNA**, where you are asked to print out the counts of each nucleotide in a specific format. Run:\n",
    "\n",
    "`rosalind info DNA`\n",
    "\n",
    "to read a description of the problem and see an example input and expected output. To save the python template at the end of the info, run the following:\n",
    "\n",
    "`rosalind info DNA > rosalind_DNA_template.py`\n",
    "\n",
    "You can then edit the script `rosalind_DNA_template.py` to complete the task. When you think your script is ready to test, run:\n",
    "\n",
    "`rosalind submit DNA rosalind_DNA_template.py`\n",
    "\n",
    "The program will show you how your script output compares with the expected output.\n",
    "\n",
    "Submit your python script to Blackboard for credit."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using Bowtie 2\n",
    "\n",
    "Today we will learn how to use the Bowtie 2 aligner. As we discussed in class, aligners use indices to efficiently find the region where a read aligns.  We will use Bowtie 2 to build an index, and then align reads to that index.  You will learn how to build indices on genomes and how to align reads to those genomes.  You will also learn how to view the alignment results using samtools. This lab is based on the tutorial for bowtie2, on the bowtie website.\n",
    "\n",
    "**For this lab you will need to be logged into SCC!** You will execute commands in your shell as they are written in this document. Remember, you must run the commands that are shown in the order they are shown or they will fail. \n",
    "\n",
    "To list all available bowtie versions on SCC:\n",
    "\n",
    "`module avail bowtie`\n",
    "\n",
    "To load the Bowtie2 module on SCC: \n",
    "module load \n",
    "\n",
    "`module load bowtie2`\n",
    "\n",
    "\n",
    "## Part A: Indices\n",
    "\n",
    "Last lab, we learned about dictionaries in Python. An index is similar to a dictionary, since we can use it to look up the location of a query sequence in a genome. Indices are used by Bowtie for memory compression and to speed up searching through the genome.  Let’s create and initialize an index on the genome of a lambda phage.\n",
    "\n",
    "1. Make a directory for lab10 in your own folder on `/projectnb/bf527/`.\n",
    "\n",
    "1. Copy the content of this lab from `/project/bf527/jupyter/BF527_Lab10_Content/` to your lab folder. \n",
    "\n",
    "   `cp -r /project/bf527/jupyter/BF527_Lab10_Content ./`\n",
    "\n",
    "1. You will need the `lambda_phage.zip` compressed files. This is a zip archive, which you can unzip with:\n",
    "\n",
    "  `unzip lambda_phage.zip`\n",
    "   \n",
    "1. There should now be a new directory in the directory you are in named `lambda_phage_example`. Use cd to enter this directory.\n",
    "1. In this new directory, you should see the following directories after running `ls`:\n",
    "\n",
    "   `index  reads  reference`\n",
    "   \n",
    "1. Type `ls index`. You should see that the `index` directory is empty. We will now build the index in the `index` directory using the following command:\n",
    "\n",
    "   `bowtie2-build reference/lambda_virus.fa index/lambda_virus`\n",
    "   \n",
    "   What this command is doing is using the program `bowtie2-build` to build an index named `lambda_virus` from the fasta file `reference/lambda_virus.fa` into the `index` directory. \n",
    "   \n",
    "1. Type `ls index`. You should see a list of files that start with the name `lambda_virus` and have various endings. These files make up your new index! To check to make sure our index was created properly, run the `bowtie2-inspect` program to print out the names of the sequences in the index:\n",
    "\n",
    "   `bowtie2-inspect -n index/lambda_virus`\n",
    "   \n",
    "   **Copy the output of the `bowtie2-inspect` command in the box below**"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gi|9626243|ref|NC_001416.1| Enterobacteria phage lambda, complete genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part B: Aligning reads\n",
    "\n",
    "Now that we have an index, we can use the index to align reads against our genome. We will stay in the `lambda_phage_example` directory for the following steps.\n",
    "\n",
    "1. Reads generated by high-throughput sequencing technologies are often provided in [FASTQ format](https://en.wikipedia.org/wiki/FASTQ_format). There are example FASTQ formatted files with short reads from the lambda phage genome in the archive you uploaded to scc. Briefly look at the format of one of these files using the following command:\n",
    "\n",
    "   `head reads/reads_1.fq`\n",
    "   \n",
    "   You don't need to understand all the contents of this file for this lab, but if you are curious the [FASTQ Wikipedia article](https://en.wikipedia.org/wiki/FASTQ_format) is a good reference.\n",
    "   \n",
    "1. We are now going to align the reads from this file against the reference we just built. Run this command:\n",
    "\n",
    "   `bowtie2 -x index/lambda_virus -U reads/reads_1.fq -S lp1.sam`\n",
    "   \n",
    "   You should see some output from running this command. The `-x` option tells bowtie the path to the index to use. `-U` tells it to use unpaired reads, and gives the name of the file the reads are in, and `-S lp1.sam` tells it to write the alignments to the file `lp1.sam` in [SAM format](https://samtools.github.io/hts-specs/SAMv1.pdf).\n",
    "   \n",
    "1. Run the command `head lp1.sam` to see the beginning of the file. This file is in SAM format - we will learn how to interpret SAM files later in the lab.\n",
    "\n",
    "1. There are two types of sequencing reads - single-end and paired-end. In the sequencing sample prep protocol, the DNA to be sequenced is sheared into short fragments, usually of 300-500nt. Single-end sequencing produces the sequence of only one end these fragments, while paired-end sequencing produces the sequence from both ends of the fragments. The sequences from paired-end datasets are stored in two different files that have the same number of reads in each, where the first read in each file form a mated pair of sequences originating from the same fragment, the second read in each file form a mated pair for another fragment, and so on.\n",
    "   \n",
    "   The file `reads_1.fq` from above is actually one of the files from a paired-end sequencing experiment, and an additional file `reads_2.fq` is also in the examples directory. Use the following to read the usage information for `bowtie2`:\n",
    "\n",
    "   `bowtie2 -h | less`\n",
    "\n",
    "    **Construct the correct command line command to run bowtie2 in paired-end mode using the files `reads_1.fq` and `reads_2.fq`, and output the alignments to `lp2.sam`.**"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "bowtie2 -x index/lambda_virus -1 reads/reads_1.fq -2 reads/reads_2.fq -S lp2.sam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part C: Analyzing SAM files\n",
    "\n",
    "Now that we’ve generated some SAM files, we need to be able to view and interpret our results. From the format description:\n",
    "\n",
    "> SAM stands for Sequence Alignment/Map format. It is a TAB-delimited text format consisting of a header section, which is optional, and an alignment section. If present, the header must be prior to the alignments. Header lines start with `@`, while alignment lines do not. Each alignment line has 11 mandatory fields for essential alignment information such as mapping position, and variable number of optional fields for flexible or aligner specific information.\n",
    "    \n",
    "We will be using the samtools toolset to process the SAM files generated by Bowtie.\n",
    "\n",
    "To list all available samtools versions on SCC:\n",
    "\n",
    "`module avail samtools`\n",
    "\n",
    "To load the samtools module on SCC:\n",
    "\n",
    "`module load samtools`\n",
    "\n",
    "1. The first thing we will do is convert the SAM file into a BAM file. A BAM file is equivalent to a SAM file, but uses a more efficient indexed and compressed binary format.\n",
    "\n",
    "   `samtools view -bS lp2.sam > lp2.bam`\n",
    " \n",
    "1. We will turn this file into a sorted BAM file. Sorted BAM files make it easy to perform a variant analysis, which is what we will be doing next:\n",
    "\n",
    "   `samtools sort lp2.bam -o lp2.sorted.bam`\n",
    " \n",
    "1. Now we will use samtools to generate a VCF file of the variants that were found. A VCF file is the standard file format used to describe variants found during sequence analysis. Make sure that you type the following command all on one line.\n",
    "\n",
    "   `samtools mpileup -uf reference/lambda_virus.fa lp2.sorted.bam | bcftools call -O b -vc --ploidy 1 > lp2.raw.bcf`\n",
    " \n",
    "1. Finally, we will use bcftools to view the VCF file:\n",
    "\n",
    "   `bcftools view lp2.raw.bcf`\n",
    "   \n",
    "**Copy the first three lines of the output into the following box.** You don't need to interpret the output of this VCF file."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "##fileformat=VCFv4.2\n",
    "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
    "##samtoolsVersion=1.10+htslib-1.10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lab 10 Task\n",
    "\n",
    " Use `bowtie2-build` to create an index for a genome, and then `bowtie2` to align reads to that genome. Then, using `samtools`, view the alignments and pass to `bcftool` to call variants. Use this information to determine if there are indels in the reads that you have aligned. \n",
    " \n",
    "### Objectives:\n",
    "\n",
    "1. In `BF527_Lab10_Content`, you should have the files `ecolicompletegenome.fasta` and `query_seqs.fasta` (along with others) in your lab directory.\n",
    "\n",
    "2. Create a new folder named `ecoli`, and in it make an index for the E. Coli genome using `ecolicompletegenome.fasta` and `bowtie2-build`.\n",
    "\n",
    "3. Run the `bowtie2` aligner, using the query sequences in `query_seqs.fasta`, against the index you created in step 2. Hint: Use `-f` to use fasta instead of fastq\n",
    "\n",
    "4. Use `samtools` to generate a BAM file and a sorted BAM file, and use `bcftools call` to call variants. Then use `bcftools view` as above to look at the results.\n",
    "\n",
    "5. What did you find? Paste the output alignment as well as a brief statement into the box below. "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "##fileformat=VCFv4.2\n",
    "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
    "##samtoolsVersion=1.10+htslib-1.10\n",
    "##samtoolsCommand=samtools mpileup -uf ecolicompletegenome.fasta al1.sorted.bam\n",
    "##reference=file://ecolicompletegenome.fasta\n",
    "##contig=<ID=gi|26245917|ref|NC_004431.1|,length=5231428>\n",
    "##ALT=<ID=*,Description=\"Represents allele(s) other than observed.\">\n",
    "##INFO=<ID=INDEL,Number=0,Type=Flag,Description=\"Indicates that the variant is an INDEL.\">\n",
    "##INFO=<ID=IDV,Number=1,Type=Integer,Description=\"Maximum number of reads supporting an indel\">\n",
    "##INFO=<ID=IMF,Number=1,Type=Float,Description=\"Maximum fraction of reads supporting an indel\">\n",
    "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Raw read depth\">\n",
    "##INFO=<ID=VDB,Number=1,Type=Float,Description=\"Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)\",Version=\"3\">\n",
    "##INFO=<ID=RPB,Number=1,Type=Float,Description=\"Mann-Whitney U test of Read Position Bias (bigger is better)\">\n",
    "##INFO=<ID=MQB,Number=1,Type=Float,Description=\"Mann-Whitney U test of Mapping Quality Bias (bigger is better)\">\n",
    "##INFO=<ID=BQB,Number=1,Type=Float,Description=\"Mann-Whitney U test of Base Quality Bias (bigger is better)\">\n",
    "##INFO=<ID=MQSB,Number=1,Type=Float,Description=\"Mann-Whitney U test of Mapping Quality vs Strand Bias (bigger is better)\">\n",
    "##INFO=<ID=SGB,Number=1,Type=Float,Description=\"Segregation based metric.\">\n",
    "##INFO=<ID=MQ0F,Number=1,Type=Float,Description=\"Fraction of MQ0 reads (smaller is better)\">\n",
    "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"List of Phred-scaled genotype likelihoods\">\n",
    "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
    "##INFO=<ID=AF1,Number=1,Type=Float,Description=\"Max-likelihood estimate of the first ALT allele frequency (assuming HWE)\">\n",
    "##INFO=<ID=AF2,Number=1,Type=Float,Description=\"Max-likelihood estimate of the first and second group ALT allele frequency (assuming HWE)\">\n",
    "##INFO=<ID=AC1,Number=1,Type=Float,Description=\"Max-likelihood estimate of the first ALT allele count (no HWE assumption)\">\n",
    "##INFO=<ID=MQ,Number=1,Type=Integer,Description=\"Root-mean-square mapping quality of covering reads\">\n",
    "##INFO=<ID=FQ,Number=1,Type=Float,Description=\"Phred probability of all samples being the same\">\n",
    "##INFO=<ID=PV4,Number=4,Type=Float,Description=\"P-values for strand bias, baseQ bias, mapQ bias and tail distance bias\">\n",
    "##INFO=<ID=G3,Number=3,Type=Float,Description=\"ML estimate of genotype frequencies\">\n",
    "##INFO=<ID=HWE,Number=1,Type=Float,Description=\"Chi^2 based HWE test P-value based on G3\">\n",
    "##INFO=<ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
    "##bcftools_callVersion=1.16+htslib-1.16-5-g2ff03d3\n",
    "##bcftools_callCommand=call -O b -vc --ploidy 1; Date=Tue Oct 18 20:47:08 2022\n",
    "##bcftools_viewVersion=1.16+htslib-1.16-5-g2ff03d3\n",
    "##bcftools_viewCommand=view al1.raw.bcf; Date=Tue Oct 18 20:47:45 2022\n",
    "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tal1.sorted.bam\n",
    "gi|26245917|ref|NC_004431.1|\t1880\t.\tTCCC\tTCCCCCC\t4.41606\t.\tINDEL;IDV=1;IMF=1;DP=1;SGB=-0.379885;MQ0F=0;AF1=1;AC1=1;DP4=0,0,1,0;MQ=40;FQ=-999\tGT:PL\t1:40,3,0\n",
    "\n",
    "The output shows there is an alteration found in this alignment where reference(ecoli genome) is TCCC and  query seqence is TCCCCCC, the quality is 4.41606."
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
